### DISCUSSION TITLE
超级QBE查询

### DISCUSSION THREAD

**Discussion by povs2012** [babyfish-ct/jimmer#753] [Ideas]
现在的QBE查询很强大，是否可以参考jhipster的过滤字段的功能https://www.jhipster.tech/entities-filtering/，通过定义几种类型的Filter，如Filter<Long> field1、Filter<String> field2，这样field1就自动支持了eq、ne、gt、lt等查询，field2自动支持eq、ne、like等查询，这样就不需要对field做alias，可以跟原属性名保持一致

**Comment by babyfish-ct** [comment]
不能，jimmer有一个设计目标，100%强类型，不会引入基于字符串的动态匹配。

**Reply by povs2012** [reply]
我可能说的不太清楚，并不是基于字符串动态匹配，就是后端定义的是Filter，针对可比较大小的定义ComparableFilter，大概类似下面的代码，后端仍然是强类型的，跟现有的QBE应该没有本质上的区别，我刚开始了解jimmer，可能有说的不对的地方
```
public class Filter<FIELD_TYPE> implements Serializable {
    private static final long serialVersionUID = 1L;

    private FIELD_TYPE equals;
    private FIELD_TYPE notEquals;
    private List<FIELD_TYPE> in;
    private List<FIELD_TYPE> notIn;
    private Boolean specified;

   // 省略构造函数、getter、setter等
}
public class ComparableFilter<FIELD_TYPE extends Comparable<? super FIELD_TYPE>> extends Filter<FIELD_TYPE> {
    @Serial
    private static final long serialVersionUID = 1L;

    private FIELD_TYPE greaterThan;
    private FIELD_TYPE lessThan;
    private FIELD_TYPE greaterThanOrEqual;
    private FIELD_TYPE lessThanOrEqual;

   // 省略构造函数、getter、setter等
}
```

**Comment by babyfish-ct** [comment]
很深的table join和sub query没法用这种方法描述。

正是因为这种方法用于描述单表，无法动态查询针对多表，所以并未采用这种为单表设计的表达方式。

如果不想用DTO语言的Specification DTO实现复杂动态查询，直接使用DSL，以kotlin为例
```
val books = sql.executeQuery(Book::class) {
    where(table.name `eq?` nullableName)
    where(table.store.name `eq?` nullableStoreName) //动态join
    where += table.authors { //动态implicit subquery
        firstName `eq?` nullableAuthorName or
        lastName `eq?` nullableAuthorName
    }
    select(table.fetch(抓取器或DTO类型))
}
```
DSL写法是静态的，单里面的每个predicate都是动态的，predicate的有无自动控制table join或sub query的舍弃（如果更复杂的场景，多个条件都成立，table join或subquery重复了，也可以自动合并）

可见，jimmer本身的DSL为动态查询设计，而且包含table join或sub query的（多表动态查询），已经足够简单。

specification DTO仅仅为了方便人们把API暴露和SQL逻辑二合一。事实上，即便不用specification DTO，也简单到极致，**无法再简化了**。

Jimmer任何功能（查询返回，数据保存，和条件匹配）立足点都是多表系统和任意形状的的数据结构。没有任何一个功能为单表操作这种简单目的。所以，Jimmer不会提供任何有单表倾向的API。

有趣的是，Jimmer下做任意复杂的多表操作，都比其他ORM处理单表简单的多。

**Reply by povs2012** [reply]
确实这种更适合单表或很简单的多表关联查询，我大致浏览了jimmer的文档，jimmer的很多想法满足了我以前使用JPA时觉得不方便的地方，是个很好的框架，值得深入学习。

**Reply by babyfish-ct** [reply]
越是深入，越是惊喜多。欢迎入群，里面jimmer高手众多。

主要是大厂乱带节奏，整体环境太无语。一起加油，改变未来，聪明地工作，而非辛苦地工作。
